wandb/util.py

```diff
@@ -891,6 +908,12 @@ def no_retry_auth(e: Any) -> bool:
     # Don't retry bad request errors; raise immediately
     if e.response.status_code in (400, 409):
         return False
+    if e.response.status_code == 404:
+        # Don't retry 404 errors, raise 404 error if it has a non-empty error message
+        # Otherwise trickle down to the default error handling below
+        error_messages = parse_backend_error_messages(e.response)
+        if error_messages:
+            return False
```

Comment:
The only other `if` statement in this function that isn't guaranteed to exit the function is 

```python
    if hasattr(e, "exception"):
        e = e.exception
```

The inconsistency makes reading this more challenging than necessary. I think it could at least be made better by not nesting an `if` statement:

```python
if (
    e.response.status_code == 404
    and parse_backend_error_messages(e.response)
):
    return False
```

But it still takes some effort to figure out what happens for 404: the next `if` statement mentions 404... but specifically to *skip* it. Then the following `if` statement also applies to something other than the 404 status. The final outcome depends on whether `wandb.run` is truthy.

The actual logic is simple: if the status is 404, then we return any error in the response or raise a CommError about missing permissions. Translating that to Python looks like

```python
if e.response.status_code == 404:
    if parse_backend_error_messages(e.response):
        # Don't retry, and use the error message in the response.
        return False
    else:
        # Don't retry, and raise a generic permission error.
        raise _permission_error(wandb.run, e.response)
```

But see my other comment: should this also apply to 401 and 403?

Reply:
Changed the code to address every error we aren't re-trying. here is summary

- 401 `ErrUnauthorized` gives the more informational API authentication error `The API key you provided is either invalid or missing...{}{}` 
- 403 is the true `ForbiddenError` and means user doesn't have permission. This gives either `Permission denied..{wandb.run.path}` if there is a run or `It appears that you do not have permission.....{}` if no run
- 404 is `ErrNotFound` and we want to return whatever backend error there is, and if there is no error we return a more generic default `Failed to find resource. Please make sure you have the correct resource path.`

Reply:
Thank you, looks much better!